### DISCUSSION TITLE
NAudio with native aot

### DISCUSSION THREAD

**Discussion by ejazmusavi** [naudio/NAudio#1103] [General]
Is there a way i can use this amazing libraray in . net core 8 application with AOT native publish. 
any help or clue to do so please.

**Comment by markheath** [comment]
I haven't tried it, but I suspect it is not possible I'm afraid. You could try setting <IsAotCompatible>true</IsAotCompatible> and compiling and I suspect that there would be warnings from the analyzers

**Reply by edward-a** [reply]
Are there any plans to make NAudio AOT compatible? Per my understanding it needs to be converted to use ComWrappers?

**Comment by edward-a** [comment]
Probably, use GeneratedComInterface?
https://learn.microsoft.com/en-us/dotnet/standard/native-interop/comwrappers-source-generation

**Comment by BradChase2011** [comment]
We have NAudio running NativeAOT on the xbox (community - wrapped NativeAOT compile) and under windows.  NativeAOT allows for com.  We had to change a few things like the WaveHeader needed to be a struct and then the memory for it declared using AllocHGlobal and then cleared in the deconstructor (dispose).  The calls to WaveInterop had to be changed to using IntPtrs instead of the WaveHeader class.  This is because the header has to be written to from the com side and obviously that cant happen using the current code.  Specifically for the xbox running NativeAOT, we had to add a few things into the interop like "waveOutMessage" and "waveInMessage" to get the default audio device since the xbox cant declare using device ID "-1".

Minor changes really, and the same code runs under regular windows and xbox windows.

This is of course all using the WinMM portion of NAudio.  So its definitely possible and quite easy.

**Reply by markheath** [reply]
That's cool to hear. Thanks for sharing your experience.

**Comment by ghost1372** [comment]
just use:
`	  <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
`

**Comment by markheath** [comment]
Good news - Native AOT is supported in NAudio 3. NAudio.Core, NAudio.Midi, NAudio.Wasapi, NAudio.Dmo, NAudio.Alsa, NAudio.Sampler and NAudio.Sampler all set `IsAotCompatible=true`, and all the COM interop has been written to use source-generated ComWrappers / [GeneratedComInterface], so no workarounds are needed. There's an AOT smoke test app that publishes with `PublishAot=true` , and exercises various WASAPI and MF capabilities.

The main gap is WinMM (`WaveOut`/`WaveIn` etc) - but I recommend using `WasapiPlayer`/`WasapiRecorder` going forwards as WASAPI is the modern way to play/record audio on Windows.

**Reply by ghost1372** [reply]
Awesome, Thank you
